src/box/iproto.cc

```diff
@@ -1865,8 +1865,6 @@ net_send_msg(struct cmsg *m)
 	if (evio_has_fd(&con->output)) {
 		if (! ev_is_active(&con->output))
 			ev_feed_event(con->loop, &con->output, EV_WRITE);
-	} else if (iproto_connection_is_idle(con)) {
```

Comment:
It's not unnecessary check! I put it back, added a `print` to this `if` and `__builtin_trap`.
```c
else if (iproto_connection_is_idle(con)) {
               fprintf(stderr, "TRAP !!!!!!!!!!!!!!!!\n");
		__builtin_trap();
		iproto_connection_close(con);
}

Reply:
You are right. I dropped it

Reply:
Good, but now in case we close connection there is no need to
````c
	if (msg->header.type == IPROTO_SHUTDOWN)
		shutdown(msg->connection->input.fd, SHUT_RD);
```
So after iproto_connection_close(con); you can add goto msg_delete; and add label msg_delete before `iproto_msg_delete(msg);` as i think.

Reply:
Also i find strange (as i think) behaviour. I start server:
```
os.execute("rm -r *.snap *.xlog")
box.cfg {
        listen = "127.0.0.1:8080"
}
box.schema.user.grant('guest','execute,write,read','universe')
_ = box.schema.space.create('counter')
_ = box.space.counter:create_index('primary')
require('console').start()
```
And i start client:
```
net_box = require('net.box')
box.cfg {
        listen = "127.0.0.1:8081"
}
graceful_con = net_box.connect("127.0.0.1", 8080)
graceful_con.space._session_settings:update('graceful_shutdown', {{'=', 'value', true}})
require('console').start()
```

And than i stop server `os.exit()`. I see shutdown message on client side, but server failed to graceful shutdown with error: "on_shutdown triggers are timed out: not all triggers might have finished yet"
I think, that client must approve shutdown and server should stop without this error!

Reply:
It is normal behavior - server waits for IPROTO_SHUTDOWN from client, but client don't send it.

Reply:
But why client, which is tarantool instance not send it? I think that client should support graceful shutdown also, and if server send shutdown to client, client must send shutdown  in response!

Reply:
Client may want to do some work after getting IPROTO_SHUTDOWN, so it is strange to send shutdown in response automatically after getting IPROTO_SHUTDOWN from server.

Reply:
Decided to do as you write, implementation details here -  https://github.com/tarantool/tarantool/issues/5924#issuecomment-838331461

Reply:
@Korablev77 what do you think about it? Should client register function for graceful shutdown support or
this should be supported by default?

Reply:
Also it's not work for more then one connection!
```lua
graceful_con_1 = net_box.connect("/tmp/1")
graceful_con_1.space._session_settings:update('graceful_shutdown', {{'=', 'value', true}})
net_box.set_shutdown_handler(function() print("XXXX") graceful_con_1:shutdown({is_async=true}) end)
graceful_con_2 = net_box.connect("/tmp/1")
graceful_con_2.space._session_settings:update('graceful_shutdown', {{'=', 'value', true}})
net_box.set_shutdown_handler(function() print("YYYY") graceful_con_2:shutdown({is_async=true}) end)
```
Only second registered function will be called and server unable to stop gracefully

Reply:
You can do `net_box.set_shutdown_handler(function() graceful_con_1:shutdown({is_async=true}) graceful_con_2:shutdown({is_async=true}) end`

Now it is one `shutdown` handler for `net.box` and there are some problems that I don't know how to solve:

1 As you showed above - client must create a function that shutdowns all connections that support graceful shutdown
2 Handler can't send synchronous requests to the server

Reply:
Now client always send `shutdown` to the server after getting `shutdown` from it. Also added `connection:set_shutdown_handler` to do something after it(after getting `shutdown` and sending `shutdown` in response).